### PR TITLE
feat: batching state updates

### PR DESCRIPTION
**Pull request by drdelambre**
this moves the csv parser into a web worker to take parsing of large annotated CSVs out of the main UX thread
also caches query parsing as that wasm was generating a lot of overhead with so many mallocs
also the biggest change is that we now batch state updates for when queries are returned while another query is processing. react has a weird thing where it'll enqueue these state updates and run them separately, which causes a bunch of collisions

some memoization tasks snuck in here. they're net positive towards the perf, but only improved it by a little, where as batching halfed (or more) processing times. keeping them in here, but there's more that can be done to minimize the cost of updating the result / status states

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -215,6 +216,7 @@
     "use-local-storage-state": "^9.0.2",
     "uuid": "^3.2.1",
     "webpack-bundle-analyzer": "^3.6.0",
+    "worker-plugin": "^5.0.1",
```
Comment (asalem1) [inline package.json:219]: is this just for IE support or what? Like, why not use the native `Worker` api?

File: src/flows/components/FlowPipe.tsx [hunk 2]
```diff
@@ -10,23 +9,22 @@ export interface FlowPipeProps {
...
+        <Pipe
+          Context={props => {
+            const _props = {
+              ...props,
+              id,
+            }
+            return createElement(FlowPanel, _props)
+          }}
+        />
+      </PipeProvider>
```
Comment (asalem1) [inline src/flows/components/FlowPipe.tsx:24]: sanity check - why was the previous implementation being used and why the change?

File: src/flows/context/flow.query.tsx [hunk 3]
```diff
@@ -374,20 +374,21 @@ export const FlowQueryProvider: FC = ({children}) => {
...
-              setStatuses({[stage.id]: RemoteDataState.Done})
               setResult(stage.id, response)
+              setStatuses({[stage.id]: RemoteDataState.Done})
             })
             .catch(e => {
-              setStatuses({[stage.id]: RemoteDataState.Error})
               setResult(stage.id, {
                 error: e.message,
               })
+              setStatuses({[stage.id]: RemoteDataState.Error})
```
Comment (asalem1) [inline src/flows/context/flow.query.tsx:391]: not sure about the level of effort, but should these be coupled to reduce some of that enqueueing you mentioned? Like consolidate the status and result into 1 result object with 2 properties?

File: src/flows/context/pipe.tsx [hunk 4]
```diff
@@ -52,13 +54,13 @@ export const PipeProvider: FC<PipeContextProps> = ({id, children}) => {
...
+          results: result || {...DEFAULT_CONTEXT.results},
+          loading: status || RemoteDataState.NotStarted,
           readOnly: flow.readOnly,
         }}
       >
         {children}
       </PipeContext.Provider>
     )
-  }, [flow, id, results, statuses, children, updater])
+  }, [flow, id, result, status, children, updater])
```
Comment (asalem1) [inline src/flows/context/pipe.tsx:65]: 👍🏽

File: src/flows/context/results.tsx [hunk 5]
```diff
@@ -32,41 +33,85 @@ export const ResultsProvider: FC = ({children}) => {
   const [statuses, setStatuses] = useState<Hash<RemoteDataState>>({})
   const [results, setResults] = useState({...EMPTY_STATE})
 
+  const updates = useRef([])
```
Comment (asalem1) [inline src/flows/context/results.tsx:36]: Never thought i'd live to see the day

File: src/flows/context/flow.query.tsx [hunk 6]
```diff
@@ -374,20 +374,21 @@ export const FlowQueryProvider: FC = ({children}) => {
...
+
     Promise.all(
       map
         .filter(q => !!q.visual)
         .map(stage => {
           return query(stage.visual, stage.scope)
             .then(response => {
-              setStatuses({[stage.id]: RemoteDataState.Done})
               setResult(stage.id, response)
+              setStatuses({[stage.id]: RemoteDataState.Done})
```
Comment (drdelambre) [inline src/flows/context/flow.query.tsx:385]: this change gets rid of a weird flash of the old results would show before showing the new results

File: src/flows/context/results.tsx [hunk 7]
```diff
@@ -32,41 +33,85 @@ export const ResultsProvider: FC = ({children}) => {
...
+    if (rChange) {
+      setResults({...results})
+    }
+
+    if (sChange) {
+      setStatuses({...statuses})
+    }
+
+    setInc(inc + 1)
+  }, [inc])
```
Comment (asalem1) [inline src/flows/context/results.tsx:73]: this feels super wrong. I get why you're doing it, but it's inherently embedding another layer of complexity where we're now using state to manage state setting? That's at least what it feels like. I know why you're doing it, I just don't like it.

File: src/flows/context/results.tsx (hunk 5 above)
Comment (drdelambre) [inline src/flows/context/results.tsx:36]: this is similar to how y-redis handles decoupling two async event loops

File: src/flows/context/results.tsx [hunk 8]
```diff
@@ -32,41 +33,85 @@ export const ResultsProvider: FC = ({children}) => {
...
+
   const value = {
     results,
     setResult: (id: string, result: Partial<FluxResult>) => {
-      results[id] = {
-        source: '',
-        parsed: null,
-        ...results[id],
-        ...result,
+      if (!isFlagEnabled('fastFlows')) {
```
Comment (asalem1) [inline src/flows/context/results.tsx:78]: Omg this flag name with the `Run` notebook functionality

File: src/flows/pipes/MetricSelector/view.tsx [hunk 9]
```diff
@@ -35,7 +35,7 @@ const DataSource: FC<PipeProp> = ({Context}) => {
   }
 
   return (
-    <BucketProvider panel={id}>
+    <BucketProvider>
```
Comment (asalem1) [inline src/flows/pipes/MetricSelector/view.tsx:38]: 👍🏽

File: src/flows/pipes/QueryBuilder/context.tsx [hunk 10]
```diff
@@ -349,7 +355,9 @@ export const QueryBuilderProvider: FC = ({children}) => {
           })
           setCardMeta([...cardMeta])
         })
-        .catch(() => {})
+        .catch(e => {
+          console.error(e)
```
Comment (asalem1) [inline src/flows/pipes/QueryBuilder/context.tsx:359]: and possible report an event?

File: src/flows/pipes/Table/index.ts [hunk 11]
```diff
@@ -1,5 +1,5 @@
 import View from './view'
-import {parse} from '@influxdata/flux'
+import {parseQuery} from 'src/shared/contexts/query'
```
Comment (asalem1) [inline src/flows/pipes/Table/index.ts:2]: why?

**Review by asalem1 — COMMENTED** [review]
Submitting what I've got on my mind before checking out the webworker stuff

File: src/shared/contexts/csv.worker.ts [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+    }
+  }
+}
+
+const ctx: Worker = self as any
+
+ctx.addEventListener('message', msg => {
+  const [idx, csv] = msg.data
+  ctx.postMessage([idx, fromFlux(csv)])
+})
```
Comment (asalem1) [inline src/shared/contexts/csv.worker.ts:358]: so what I'm gather from this file is that the webworker can't make use of existing files or functions, or dependencies right? That's the reason for the pasta? Nothing new here right?

File: src/shared/contexts/query.tsx [hunk 13]
```diff
@@ -392,12 +407,126 @@ export const simplify = (text, vars: VariableMap = {}) => {
...
+  }
+})()
+
 export const QueryProvider: FC = ({children}) => {
   const buckets = useSelector((state: AppState) => getSortedBuckets(state))
   const bucketsLoadingState = useSelector((state: AppState) =>
     getStatus(state, ResourceType.Buckets)
   )
-  const [pending, setPending] = useState({} as CancelMap)
+  const pending = useRef({} as CancelMap)
```
Comment (asalem1) [inline src/shared/contexts/query.tsx:529]: 🤯

**Comment by ui-ci-bot-influx** [comment]
Metric | Score | Level
------------ | ------------- | ------------- 
performance | 60 | :red_circle:
accessibility | 96 | :green_circle:
seo | 90 | :green_circle:
pwa | null | :hankey:
best-practices | 100 | :green_circle:


[Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?lower=now%28%29%20-%202d&vars%5Blighthouse_score_branch%5D=2983)

File: src/flows/components/FlowPipe.tsx (hunk 2 above)
Comment (drdelambre) [reply src/flows/components/FlowPipe.tsx:24]: i think it's more of an oversight of when we added the "pass the container as a prop" idea. turns out that oversight cost us 15ms over a large number of results coming in

File: src/flows/pipes/Table/index.ts (hunk 11 above)
Comment (drdelambre) [reply src/flows/pipes/Table/index.ts:2]: good catch. needs to be wrapped in a flag

File: src/shared/contexts/csv.worker.ts (hunk 12 above)
Comment (drdelambre) [reply src/shared/contexts/csv.worker.ts:358]: nothing new, except we can't have anything more than a plain old javascript object being passed back and forth (had to remove simpeTable). we couldn't import a giraffe function directly because of how the exports are exposed in the npm module.

File: src/shared/contexts/csv.worker.ts [hunk 14]
```diff
@@ -0,0 +1,358 @@
+import {csvParse, csvParseRows} from 'd3-dsv'
+
+interface FromFluxResult {
+  error?: Error
+  // The single parsed `Table`
+  table: Table
+
+  // The union of unique group keys from all input Flux tables
+  fluxGroupKeyUnion: string[]
+}
```
[the remark this answers is not in the document]
Comment (drdelambre) [reply src/shared/contexts/csv.worker.ts:10]: too many, and after all of this is migrated over... it'll just be the one in src/types/flows
